sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -127,6 +127,176 @@ case class MapKeys(child: Expression)
+    val genericInternalRow = classOf[GenericInternalRow].getName
+    val arrVals = ctx.freshName("arrVals")
+    val arrCardinality = ctx.freshName("arrCardinality")
+    val biggestCardinality = ctx.freshName("biggestCardinality")
+    val storedArrTypes = ctx.freshName("storedArrTypes")
+    val returnNull = ctx.freshName("returnNull")
+    val evals = children.map(_.genCode(ctx))
+
+    val inputs = evals.zipWithIndex.map { case (eval, index) =>
+      s"""
+        |${eval.code}
+        |if (!${eval.isNull}) {
+        |  $arrVals[$index] = ${eval.value};
+        |  $arrCardinality[$index] = ${eval.value}.numElements();
+        |} else {
+        |  $arrVals[$index] = null;
+        |  $arrCardinality[$index] = 0;
+        |  $returnNull[0] = true;
+        |}
+        |$storedArrTypes[$index] = "${arrayElementTypes(index)}";
+        |$biggestCardinality = Math.max($biggestCardinality, $arrCardinality[$index]);
+      """.stripMargin
+    }
+
+    val inputsSplitted = ctx.splitExpressions(
+      expressions = inputs,
+      funcName = "getInputAndCardinality",
+      returnType = "int",
+      makeSplitFunction = body =>
+        s"""
+          |$body
+          |return $biggestCardinality;
+        """.stripMargin,
+      foldFunctions = _.map(funcCall => s"$biggestCardinality = $funcCall;").mkString("\n"),
+      arguments =
+        ("ArrayData[]", arrVals) ::
+        ("int[]", arrCardinality) ::
+        ("String[]", storedArrTypes) ::
+        ("int", biggestCardinality) ::
+        ("boolean[]", returnNull) :: Nil)
+
+    val myobject = ctx.freshName("myobject")
+    val j = ctx.freshName("j")
+    val i = ctx.freshName("i")
+    val args = ctx.freshName("args")
+
+    val cases = arrayElementTypes.distinct.map { elementType =>
+      val getArrValsItem = CodeGenerator.getValue(s"$arrVals[$j]", elementType, i)
+      s"""
+        |case "${elementType}":
+        |  $myobject[$j] = $getArrValsItem;
+        |  break;
+      """.stripMargin
+    }
+
+    ev.copy(s"""
+      |ArrayData[] $arrVals = new ArrayData[$numberOfArrays];
+      |int[] $arrCardinality = new int[$numberOfArrays];
+      |int $biggestCardinality = 0;
+      |String[] $storedArrTypes = new String[$numberOfArrays];
+      |boolean[] $returnNull = new boolean[1];
+      |$returnNull[0] = false;
+      |$inputsSplitted
+      |${CodeGenerator.javaType(dataType)} ${ev.value};
+      |Object[] $args = new Object[$biggestCardinality];
+      |for (int $i = 0; $i < $biggestCardinality; $i ++) {
+      |  Object[] $myobject = new Object[$numberOfArrays];
+      |  for (int $j = 0; $j < $numberOfArrays; $j ++) {
+      |    if ($arrVals[$j] != null && $arrCardinality[$j] > $i && !$arrVals[$j].isNullAt($i)) {
+      |      switch ($storedArrTypes[$j]) {
+      |        ${cases.mkString("\n")}
+      |       default:
+      |         break;
+      |      }
+      |    } else {
+      |      $myobject[$j] = null;
+      |    }
+      |  }
```

Comment:
On second thought, I think we shouldn't use a (inner) loop to extract values from arrays because:

- I'm not sure the string comparison of data types works properly for complex types.
- Even if it works, we can't split the switch statement I suggested when the statement becomes bigger.
- When going back to if-else statement, a bunch of if-else evaluations for each array would cause performance issue.
- We can know the exact accessing method for each array.

WDYT? @kiszk @mgaido91 @mn-mikke

Reply:
If we can remove `storedArrTypes` as I pointed out at https://github.com/apache/spark/pull/21045#discussion_r191126615, can we do eliminate `switch-case`? It can make generated code simple.

Reply:
@DylanGuedes I think that @ueshin 's suggestion is that we create a new loop to create an array to store `CodeGenerator.getValue(s"$arrVals[$j]", elementType, i)`. Then, we execute the existing nested loop without executing `getValues` at the innermost. Am I correct?

Reply:
I don't get it, could you detail more? I'm calling/eval'ing these multiple `getValues` before reaching `ev.copy` (since it is Scala code) and just injecting its result in the for - I'm not generating multiples `getValues` for each iteration of the `for` loop - they were already eval'd and I'm just injecting them.

Just in case, I reverted to use `if` instead of `switch-case`. The problem is that I can't guess which of these `getValues` to use without hardcodding them.